### ISSUE TITLE
Unable to set VideoCodecType

### ISSUE THREAD

**Issue by Komal-Khairnar** [AgoraIO-Extensions/Agora-Flutter-SDK#332]
How to set the codec for the video using VideoCodecType?

**Comment by LichKing-2234** [comment]
On Flutter Web?

**Comment by kazy-dev** [comment]
I am implementing a regular one-on-one video calling application.
I too would like to know how to set the codec.

The documentation says to call startRtmpStreamWithTranscoding() when using liveTranscoding, but the codec-related arguments I can specify are
VideoCodecTypeForStream videoCodecType
and there are only two definitions

https://github.com/AgoraIO/Agora-Flutter-SDK/blob/d0d6fad1f1177460c079059675fe5d2f15bcc2a2/example/lib/examples/advanced/rtmp_streaming/rtmp_streaming.dart#L129-L187

```
final liveTranscoding = LiveTranscoding(
      transcodingUsers,
      width: width,
      height: height,
      videoBitrate: 400,
      videoCodecProfile: VideoCodecProfileType.High,
      videoCodecType: ,
      videoGop: 30,
      videoFramerate: VideoFrameRate.Fps24,
      lowLatency: false,
      audioSampleRate: AudioSampleRateType.Type44100,
      audioBitrate: 48,
      audioChannels: AudioChannel.Channel1,
      audioCodecProfile: AudioCodecProfileType.LCAAC,
    );
    await _rtcEngine?.startRtmpStreamWithTranscoding(url, transcoding);
```

```
enum VideoCodecTypeForStream {
  ///
  /// 1: (Default) H.264.
  ///
  @JsonValue(1)
  H264,

  ///
  /// 2: H.265.
  ///
  @JsonValue(2)
  H265,
}
```

The following enum called VideoCodecType exists, but I can't find where to specify it.
```
///
/// Video codec types.
///
///
enum VideoCodecType {
  ///
  /// Standard VP8.
  ///
  @JsonValue(1)
  VP8,

  ///
  /// Standard H.264.
  ///
  @JsonValue(2)
  H264,

  ///
  /// Enhanced VP8.
  ///
  @JsonValue(3)
  EVP,

  ///
  ///
  ///
  @JsonValue(4)
  E264,
}
```
In agora-web-sdk, codecs can be specified as follows
AgoraRTC.createClient({ mode: this.transcode, codec: 'vp8' })
Does anyone know how to set codecs like VP8 or web in Flutter or native SDK for

It seems that when using startRtmpStreamWithTranscoding, the Rtmp stream URL must be specified.
In other words, I would like to know how to set the codec when there is no unique Rtmp stream URL.

**Comment by kazy-dev** [comment]
@LichKing-2234

**Comment by kazy-dev** [comment]
> I am implementing a regular one-on-one video calling application. I too would like to know how to set the codec.
> 
> The documentation says to call startRtmpStreamWithTranscoding() when using liveTranscoding, but the codec-related arguments I can specify are VideoCodecTypeForStream videoCodecType and there are only two definitions
> 
> https://github.com/AgoraIO/Agora-Flutter-SDK/blob/d0d6fad1f1177460c079059675fe5d2f15bcc2a2/example/lib/examples/advanced/rtmp_streaming/rtmp_streaming.dart#L129-L187
> 
> ```
> final liveTranscoding = LiveTranscoding(
>       transcodingUsers,
>       width: width,
>       height: height,
>       videoBitrate: 400,
>       videoCodecProfile: VideoCodecProfileType.High,
>       videoCodecType: ,
>       videoGop: 30,
>       videoFramerate: VideoFrameRate.Fps24,
>       lowLatency: false,
>       audioSampleRate: AudioSampleRateType.Type44100,
>       audioBitrate: 48,
>       audioChannels: AudioChannel.Channel1,
>       audioCodecProfile: AudioCodecProfileType.LCAAC,
>     );
>     await _rtcEngine?.startRtmpStreamWithTranscoding(url, transcoding);
> ```
> 
> ```
> enum VideoCodecTypeForStream {
>   ///
>   /// 1: (Default) H.264.
>   ///
>   @JsonValue(1)
>   H264,
> 
>   ///
>   /// 2: H.265.
>   ///
>   @JsonValue(2)
>   H265,
> }
> ```
> 
> The following enum called VideoCodecType exists, but I can't find where to specify it.
> 
> ```
> ///
> /// Video codec types.
> ///
> ///
> enum VideoCodecType {
>   ///
>   /// Standard VP8.
>   ///
>   @JsonValue(1)
>   VP8,
> 
>   ///
>   /// Standard H.264.
>   ///
>   @JsonValue(2)
>   H264,
> 
>   ///
>   /// Enhanced VP8.
>   ///
>   @JsonValue(3)
>   EVP,
> 
>   ///
>   ///
>   ///
>   @JsonValue(4)
>   E264,
> }
> ```
> 
> In agora-web-sdk, codecs can be specified as follows AgoraRTC.createClient({ mode: this.transcode, codec: 'vp8' }) Does anyone know how to set codecs like VP8 or web in Flutter or native SDK for
> 
> It seems that when using startRtmpStreamWithTranscoding, the Rtmp stream URL must be specified. In other words, I would like to know how to set the codec when there is no unique Rtmp stream URL.

Since agora_rtc_engine 6.0.0.
`Future<void> setVideoEncoderConfiguration(VideoEncoderConfiguration config);` allows to specify `codecType`.
This `codecType` has the following types.
```
/// Video codec types.
@JsonEnum(alwaysCreate: true)
enum VideoCodecType {
  /// @nodoc
  @JsonValue(0)
  videoCodecNone,

  /// 1: Standard VP8.
  @JsonValue(1)
  videoCodecVp8,

  /// 2: Standard H.264.
  @JsonValue(2)
  videoCodecH264,

  /// 3: Standard H.265.
  @JsonValue(3)
  videoCodecH265,

  /// 6: Generic.This type is used for transmitting raw video data, such as encrypted video frames. The SDK returns this type of video frames in callbacks, and you need to decode and render the frames yourself.
  @JsonValue(6)
  videoCodecGeneric,

  /// @nodoc
  @JsonValue(7)
  videoCodecGenericH264,

  /// @nodoc
  @JsonValue(12)
  videoCodecAv1,

  /// @nodoc
  @JsonValue(13)
  videoCodecVp9,

  /// 20: Generic JPEG.This type consumes minimum computing resources and applies to IoT devices.
  @JsonValue(20)
  videoCodecGenericJpeg,
}
```
You can specify the following.
```
final videoConfig = VideoEncoderConfiguration(
  dimensions: VideoDimensions(width: 640, height: 480),
  frameRate: VideoFrameRate.fps15,
  orientationMode: VideoOutputOrientationMode.FixedPortrait,
  encoderMirrorMode: VideoMirrorMode.Disabled,
  minBitrate: 0,
  degradationPreference: DegradationPreference.MaintainFramerate,
  mirrorMode: VideoMirrorMode.Disabled,
  bitrateAdjustmentType: BitrateAdjustmentType.Dynamic,
  videoCodecType: VideoCodecType.VP8, // Set the video codec to VP8
);

await _engine.setVideoEncoderConfiguration(videoConfig);
```
Resolved.
I hope this will be helpful for everyone.

@LichKing-2234 
This isssue may be closed.

**Comment by LichKing-2234** [comment]
@kazy-dev thanks for your comments!
